kubetest/gke.go

```diff
@@ -608,12 +617,136 @@ func (g *gkeDeployer) cleanupNetworkFirewalls() (int, error) {
 	return 0, nil
 }
 
+func (g *gkeDeployer) ensureNat() error {
+	if !g.nat {
+		return nil
+	}
+	if g.network == "default" {
+		return fmt.Errorf("NAT router should be set manually for the default network")
+	}
+	region, err := g.getRegion(g.region, g.zone)
+	if err != nil {
+		return fmt.Errorf("error finding region for NAT router: %v", err)
+	}
+	nat, err := g.getNatName()
+	if err != nil {
+		return fmt.Errorf("error getting unique NAT name: %v", err)
+	}
+
+	// Create this unique NAT router only if it does not exist yet.
+	if control.NoOutput(exec.Command("gcloud", "compute", "routers", "describe", nat,
+		"--project="+g.project,
+		"--region="+region,
+		"--format=value(name)")) != nil {
+		log.Printf("Couldn't describe router '%s', assuming it doesn't exist and creating it", nat)
+
+		if err := control.FinishRunning(exec.Command("gcloud", "compute", "routers", "create", nat,
+			"--project="+g.project,
+			"--network="+g.network,
+			"--region="+region)); err != nil {
+			return fmt.Errorf("error creating NAT router: %v", err)
+		}
+	}
+	// Create this unique NAT config only if it does not exist yet.
+	if control.NoOutput(exec.Command("gcloud", "compute", "routers", "nats", "describe", nat,
+		"--project="+g.project,
+		"--router="+nat,
+		"--router-region="+region,
+		"--format=value(name)")) != nil {
+		log.Printf("Couldn't describe nat '%s', assuming it doesn't exist and creating it", nat)
+		if err := control.FinishRunning(exec.Command("gcloud", "compute", "routers", "nats", "create", nat,
+			"--project="+g.project,
+			"--router="+nat,
+			"--router-region="+region,
+			"--auto-allocate-nat-external-ips",
+			"--nat-all-subnet-ip-ranges",
+			"--min-ports-per-vm="+strconv.Itoa(g.natMinPortsPerVm),
+			"--enable-logging")); err != nil {
+			return fmt.Errorf("error creating NAT config: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (g *gkeDeployer) getRegion(region, zone string) (string, error) {
+	if region != "" {
+		return region, nil
+	}
+	result, err := exec.Command("gcloud", "compute", "zones", "list",
+		"--filter=name="+zone,
+		"--format=value(region)",
+		"--project="+g.project).Output()
+	if err != nil {
+		return "", fmt.Errorf("error resolving region of %s zone: %v", zone, err)
+	}
+	return strings.TrimSuffix(string(result), "\n"), nil
+}
+
+func (g *gkeDeployer) getNatName() (string, error) {
+	if err := g.getInstanceGroups(); err != nil {
+		return "", err
+	}
+	// Create unique NAT router name based on hash
+	// from the lexically first node pool instead.
+	return "nat-router-" + g.instanceGroups[0].uniq, nil
```

Comment:
Cannot we use g.cluster for the nat name?

Reply:
Done